templates/pwm.asg.element.json

```diff
@@ -0,0 +1,843 @@
+						"Action": [
+							"ec2:DescribeTags"
+						],
+						"Effect": "Allow",
+						"Resource": "*"
+					}],
+					"Version": "2012-10-17"
+				},
+				"Roles": [{
+					"Ref": "PWMInstanceRole"
+				}]
+			},
+			"Type": "AWS::IAM::ManagedPolicy"
+		},
+		"PWMAutoScalingGroup": {
+			"CreationPolicy": {
+				"ResourceSignal": {
+					"Count": {
+						"Ref": "DesiredCapacity"
+					},
+					"Timeout": "PT25M"
+				}
+			},
+			"Properties": {
+				"DesiredCapacity": {
+					"Ref": "DesiredCapacity"
+				},
+				"HealthCheckGracePeriod": 3600,
+				"HealthCheckType": "ELB",
+				"LaunchConfigurationName": {
+					"Ref": "PWMLaunchConfig"
+				},
+				"MaxSize": {
+					"Ref": "MaxCapacity"
+				},
+				"MinSize": {
+					"Ref": "MinCapacity"
+				},
+				"NotificationConfigurations": [{
+					"NotificationTypes": [
+						"autoscaling:EC2_INSTANCE_LAUNCH",
+						"autoscaling:EC2_INSTANCE_LAUNCH_ERROR",
+						"autoscaling:EC2_INSTANCE_TERMINATE",
+						"autoscaling:EC2_INSTANCE_TERMINATE_ERROR"
+					],
+					"TopicARN": {
+						"Ref": "PWMSNSTopicARN"
+					}
+				}],
+				"Tags": [{
+					"Key": "Name",
+					"Value": {
+						"Ref": "AWS::StackName"
+					},
+					"PropagateAtLaunch": true
+				}],
+				"TargetGroupARNs": [{
+					"Ref": "PWMALBTargetGroupName"
+				}],
+				"VPCZoneIdentifier": {
+					"Ref": "PrivateSubnetIDs"
+				}
+			},
+			"Type": "AWS::AutoScaling::AutoScalingGroup",
+			"UpdatePolicy": {
+				"AutoScalingReplacingUpdate": {
+					"WillReplace": true
+				}
+			}
+		},
+		"PWMConfigInstanceS3Access": {
+			"Properties": {
+				"Description": "Policy for PWM instance role to access appropriate s3 bucket for pwm config",
+				"PolicyDocument": {
+					"Statement": [{
+						"Action": [
+							"s3:GetObject",
+							"s3:PutObject",
+							"s3:PutObjectAcl"
+						],
+						"Effect": "Allow",
+						"Resource": [{
+							"Fn::Join": [
+								"",
+								[
+									"arn:aws:s3:::",
+									{
+										"Ref": "ConfigBucketName"
+									},
+									"/*"
+								]
+							]
+						}]
+					}],
+					"Version": "2012-10-17"
+				},
+				"Roles": [{
+					"Ref": "PWMInstanceRole"
+				}]
+			},
+			"Type": "AWS::IAM::ManagedPolicy"
+		},
+		"PWMInstanceProfile": {
+			"Properties": {
+				"Path": "/",
+				"Roles": [{
+					"Ref": "PWMInstanceRole"
+				}]
+			},
+			"Type": "AWS::IAM::InstanceProfile"
+		},
+		"PWMInstanceRole": {
+			"Properties": {
+				"AssumeRolePolicyDocument": {
+					"Statement": [{
+						"Action": [
+							"sts:AssumeRole"
+						],
+						"Effect": "Allow",
+						"Principal": {
+							"Service": [
+								"ec2.amazonaws.com"
+							]
+						}
+					}],
+					"Version": "2012-10-17"
+				},
+				"Path": "/"
+			},
+			"Type": "AWS::IAM::Role"
+		},
+		"PWMLaunchConfig": {
+			"Metadata": {
+				"AWS::CloudFormation::Init": {
+					"configSets": {
+						"Install": [
+							"a-cfnsetup",
+							"b-bootstrap",
+							"c-adminusersetup",
+							"d-cloudwatchagentsetup",
```

Comment:
Recommend moving this step as early as possible so logs are captured and viewable in case of error in one of the steps that are currently executing before this...